lib/pg/instrumentation.rb

```diff
@@ -0,0 +1,158 @@
+            ::PG::Instrumentation.tags['peer.address'] = "pg://#{host}:#{port}"
+
+            operation_name = 'pg.initialize'
+
+            span = ::PG::Instrumentation.tracer.start_span(operation_name, tags: ::PG::Instrumentation.tags)
+
+            initialize_original(*args)
+
+            rescue => error
+              log_error(span, error)
+          end
+
+          def exec(*args)
+            return '' if args.empty?
+
+            tags = ::PG::Instrumentation.tags.dup
+
+            sql = args.first.to_s[0, 1024]
+            tags['db.statement'] = sql
+
+            default_op_name = 'pg.query'
+            operation_name = get_operation_name(sql, default_op_name)
+
+            span = ::PG::Instrumentation.tracer.start_span(operation_name, tags: tags)
+            exec_original(*args)
+
+          rescue => error
+            log_error(span, error)
+          end
+
+          def exec_params(*args)
+            return '' if args.empty?
+
+            tags = ::PG::Instrumentation.tags.dup
+
+            sql = args.first.to_s[0, 1024]
+            tags['db.statement'] = sql
+
+            default_op_name = 'pg.query'
+            operation_name = get_operation_name(sql, default_op_name)
+
+            span = ::PG::Instrumentation.tracer.start_span(operation_name, tags: tags)
+
+            exec_params_original(*args)
+
+            rescue => error
+              log_error(span, error)
+          end
+
+          def prepare(*args)
+            return '' if args.empty?
+
+            tags = ::PG::Instrumentation.tags.dup
+
+            sql = args[1].to_str[0, 1024]
+            tags['db.statement'] = sql
+            tags['prepared.statement.name'] = args.first
+
+            default_op_name = 'pg.prepare_statement'
+            operation_name = get_operation_name(sql, default_op_name)
+
+            span = ::PG::Instrumentation.tracer.start_span(operation_name, tags: tags)
+
+            prepare_original(*args)
+
+            rescue => error
+              log_error(span, error)
+          end
+
+          def exec_prepared(*args)
+            return '' if args.empty?
+
+            tags = ::PG::Instrumentation.tags.dup
+            tags['prepared.statement.name'] = args.first
+
+            operation_name = 'pg.exec_prepared_statement'
+            span = ::PG::Instrumentation.tracer.start_span(operation_name, tags: tags)
+
+            exec_prepared_original(*args)
+
+          rescue => error
+            log_error(span, error)
+          end
+
+          # Helper functions
+          def get_operation_name(sql, default)
+            sql_split = sql.split(' ')
+            candidate = sql_split[0].upcase if sql_split.length > 1
+            return candidate if !candidate.nil? && !candidate.empty?
+
+            return default
+          rescue
+          end
+
+          def log_error(span, error)
+            span ||= ::PG::Instrumentation.tracer.start_span('error_span')
```

Comment:
In what cases would we need to create a span?  If it's a bug in instrumentation we wouldn't want that leaking into user's traces right?